dali/pipeline/operators/python_function/dltensor_function.h

```diff
@@ -0,0 +1,126 @@
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_PYTHON_FUNCTION_DLTENSOR_FUNCTION_H_
+#define DALI_PIPELINE_OPERATORS_PYTHON_FUNCTION_DLTENSOR_FUNCTION_H_
+
+#include "dali/pipeline/operators/python_function/python_function.h"
+// All python headers must be included before std headers due to macro redefinition error
+#include <vector>  // NOLINT
+#include "dali/pipeline/operators/python_function/util/copy_with_stride.h"
+
+namespace dali {
+
+namespace detail {
+
+template <typename Backend>
+constexpr DLDeviceType Backend2DLDevice() {
+  if (std::is_same<Backend, CPUBackend>::value) {
+    return kDLCPU;
+  } else {
+    return kDLGPU;
+  }
+}
+
+template <typename Backend>
+std::vector<DLMTensorPtr> CastToDLTensorList(py::list &list, Index exp_size, Index out_idx) {
+  DALI_ENFORCE(list.size() == static_cast<size_t>(exp_size),
+      "Function called by DLTensorPythonFunction returned tensor list of wrong size at idx "
+      + std::to_string(out_idx) + ". Returned list is of a size "
+      + std::to_string(exp_size) + " and should be of size " + std::to_string(exp_size));
+  std::vector<DLMTensorPtr> result;
+  result.reserve(exp_size);
+  if (exp_size) {
+    DALI_ENFORCE(py::capsule::check_(list[0]),
+                 "Function called by DLTensorPythonFunction "
+                 "should return a list of DLPack tensors.");
+    auto caps = py::cast<py::capsule>(list[0]);
+    result.push_back(DLMTensorPtrFromCapsule(caps));
+    DALI_ENFORCE(result[0]->dl_tensor.ctx.device_type == Backend2DLDevice<Backend>(),
+        "Wrong output backend");
+    auto dtype = result[0]->dl_tensor.dtype;
+    for (Index i = 1; i < exp_size; ++i) {
+      auto caps = py::cast<py::capsule>(list[i]);
+      result.push_back(DLMTensorPtrFromCapsule(caps));
+      DALI_ENFORCE(result[i]->dl_tensor.ctx.device_type == Backend2DLDevice<Backend>(),
+                   "Wrong output backend.");
+      DALI_ENFORCE(DLToDALIType(result[i]->dl_tensor.dtype) == DLToDALIType(dtype),
+          "Output DLPack tensor list should have consistent data type.");
+    }
+  }
+  return result;
+}
+
+template <typename Backend>
+void CopyDlTensor(void *out_data, DLMTensorPtr &dlm_tensor_ptr) {
+  auto &dl_tensor = dlm_tensor_ptr->dl_tensor;
+  auto item_size = dl_tensor.dtype.bits / 8;
+  std::vector<Index> strides(dl_tensor.ndim);
+  if (dl_tensor.strides) {
+    for (Index i = 0; i < dl_tensor.ndim; ++i) strides[i] = dl_tensor.strides[i] * item_size;
+  } else {
+    strides.back() = item_size;
+    for (Index i = dl_tensor.ndim - 2; i >= 0; --i) {
+      strides[i] = strides[i + 1] * dl_tensor.shape[i + 1];
+    }
+  }
+  CopyWithStride<Backend>(out_data, dl_tensor.data, strides.data(),
+                          dl_tensor.shape, dl_tensor.ndim, item_size);
+}
+
+template <typename Backend>
+py::list PrepareInputs(workspace_t<Backend> &ws);
+
+template <typename Backend>
+void CopyOutputs(workspace_t<Backend> &ws, py::tuple &output);
```

Comment:
Please rename this function - the parameter types don't require or even suggest that we're dealing with DLPack here.

Reply:
1. Why do you return by output parameter here and by return value in PrepareInputs?
2. Personal preference - I think that "memcpy" argument order (output first) is easier to read.

Reply:
This `is` the "memcpy" argument order, because data is copied from output tuple, to DALI tensors, which are in the workspace. I changed the name of the `output` argument to `return_tuple`, I hope it's less confusing.